### ISSUE TITLE
Multiprocess debugging

### ISSUE THREAD

**Issue by daniel5151** [daniel5151/gdbstub#124]
`gdbstub` already implements protocol-level multiprocess extensions "under-the-hood", and simply hard-codes a fake PID in single/multi-threaded mode (much the same way it hard-codes a fake TID in single-threaded mode).

Adding true multiprocess debugging will most likely involve adding a new `target::ext::base::multiprocess` API set, along with doing the requisite plumbing to track and report the current PID.

In addition, we'd want to add a new in-tree `armv4t_multiprocess` example, which wouldn't _actually_ spin up multiple processes, but would instead spin up multiple multi-core `armv4t` emulator instances, just so there's an in-tree example of this stuff working.

Why API-breaking? Unfortunately, I neglected to mark the `BaseOps` enum as `non_exhaustive`, which means implementing this will be a breaking change. Aside from that, this likely could've been entirely non-breaking...

**Comment by xobs** [comment]
I think the only thing that needs to be implemented is `info os processes`:

```
(gdb) info os processes
[remote] Sending packet: $qXfer:osdata:read:processes:0,1000#fa
[remote] Packet received: l<osdata type="processes">\n<item><column name="pid">1</column><column name="user">root</column><column name="command">/init </column><column name="cores">6,15</column></item><item><column name="pid">4</column><column name="user">root</column><column name="command">plan9 --control-socket 5 --log-level 4 --server-fd 6 --pipe-fd 8 --log-truncate </column><column name="cores">5,19</column></item><item><column name="pid">7</column><column name="user">root</column><column name="command">/init </column><column nam [2029 bytes omitted]
pid        user       command    cores
1          root       /init      6,15
4          root       plan9 --control-socket 5 --log-level 4 --server-fd 6 --pipe-fd 8 --log-truncate 5,19
7          root       /init      16
8          root       /init      2
9          user       -bash      23
10         root       /init      11
11         root       /init      11
12         user       -bash      13
468        user       dbus-launch --autolaunch 857863f4cf286a4142beec6a62a03f1f --binary-syntax --close-stderr 20
469        user       /usr/bin/dbus-daemon --syslog-only --fork --print-pid 5 --print-address 7 --session 23
499        user       gdb        0,1,2,4,5,6,8,9,10,12,13,14,15,16,17,18,20,21,22,23
525        root       /init      17
526        root       /init      19
527        user       -bash      21
544        user       gdbserver --multi localhost:3454 14
(gdb)
```

This would then be followed by `vAttach` with a `Pid` that would be interpreted by the kernel:

```
(gdb) attach 527
Attaching to program: /usr/bin/sleep, process 527
[remote] Sending packet: $vAttach;20f#ce
[remote] Packet received: T0006:a0faeb5ce87f0000;07:30086b64ff7f0000;10:f418dc5ce87f0000;thread:p20f.20f;core:15;
[remote] packet_ok: Packet vAttach (attach) is supported
[remote] Sending packet: $qXfer:exec-file:read:20f:0,1000#e1
[remote] Packet received: l/usr/bin/bash
[remote] Sending packet: $vFile:setfs:0#bf
[remote] Packet received: F0
[remote] packet_ok: Packet vFile:setfs (hostio-setfs) is supported
[remote] Sending packet: $vFile:open:6a7573742070726f62696e67,0,1c0#ed
[remote] Packet received: F-1,2
[remote] packet_ok: Packet vFile:open (hostio-open) is supported
[remote] Sending packet: $vFile:setfs:20f#57
[remote] Packet received: F0
[remote] Sending packet: $vFile:open:2f7573722f62696e2f62617368,0,0#f4
[remote] Packet received: F5
[remote] remote_hostio_pread: readahead cache miss 1
[remote] Sending packet: $vFile:pread:5,47ff,0#6a
[remote] Packet received: F4756;\177ELF\002\001\00.........
```

Since `vAttach` is already supported, the only thing that would be needed is `info os processes`, which should be incredibly simple to implement -- just add `:osdata` to the configuration listing and respond to `w $qXfer:osdata:read:processes:0,1000#fa` with the list.

**Comment by daniel5151** [comment]
Unfortunately, I don't think it's quite that simple.

While implementing `info os processes` would be a good QOL feature (it'd give you "first class" support for listing procs in GDB vs. you having to hand-roll a custom `monitor` command), I doubt it'll affect the GDB client's internal bookkeeping when it comes to things like reading/writing memory/registers, switching between procs, etc...

The big feature in `gdbstub` is the fact that there's lots of code that either _ignores_ or _fakes_ PID values:

A great example of this is something like this:

https://github.com/daniel5151/gdbstub/blob/85ff370e018c031bc88438123bf2766c2c806d63/src/stub/core_impl/base.rs#L345-L351

See how `gdbstub` uses `FAKE_PID` here as part of it's response?

That means if you tried to `vAttach` to a process that had a PID values that isn't `FAKE_PID` (i.e: `1`), the GDB client would get _very confused_ when the `gdbstub` would begin sending back data related to an unrelated PID.

Similarly, there's code like this as well:

https://github.com/daniel5151/gdbstub/blob/85ff370e018c031bc88438123bf2766c2c806d63/src/stub/core_impl.rs#L125-L126

Namely: `gdbstub` _should_ be tracking `current_mem_tid_pid`, and `current_resume_tid_pid`, but instead, the code that sets/reads from these variables either discards the `pid` the GDB client sends, or sends back a FAKE_PID.

...things like that are why the _correct_ solution here would be to plumb through a whole new `target::ext::base::multiprocess` interface, which - to be clear - would be a straight mirror of the existing `target::ext::base::multithread` API, except instead of using `Tid` in all the APIs, it'd use `(Pid, Tid)` instead.

**Comment by xobs** [comment]
I see -- I thought "multiprocess" meant "inferior support", which seems to be mostly managed by the GDB client itself. That is, when debugging multiple concurrent processes, it's the job of the GDB client to launch multiple processes or initiate connections via the bridge, and it's not the job of the bridge to make those connections.

Accordingly, there is no way to get a list of running processes with `inferior` support -- the only way to do it is `info os processes`.

I suppose `target::ext::base::multiprocess` would be distinct from `inferior` support?

**Comment by daniel5151** [comment]
> I see -- I thought "multiprocess" meant "inferior support", which seems to be mostly managed by the GDB client itself. That is, when debugging multiple concurrent processes, it's the job of the GDB client to launch multiple processes or initiate connections via the bridge, and it's not the job of the bridge to make those connections.

Unless I'm misunderstanding something, I don't believe there's much distinction at the GDB RSP layer between "managed" inferiors vs. "attached" inferiors, aside from some client-side book-keeping regarding whether the process should be terminated on disconnect, vs. automatically resumed, among other similar trivialities. A process is a process, regardless if the GDB client has told the remote stub to spin it up itself vs. asking the remote host for a list of processes + attaching to some existing one.

> I suppose `target::ext::base::multiprocess` would be distinct from `inferior` support?

I think a better way to think about it is in the sense that `gdbstub` has _always_ supported a _single_ inferior, but without a `target::ext::base::multiprocess` API, is unable to supporting _multiple_ inferiors (both in the sense of simultaneous debugging, and in the sense of switching between different inferiors on-the-fly).

**Comment by xobs** [comment]
With https://github.com/daniel5151/gdbstub/pull/129 I'm able to switch processes. The hardest part is that `FAKE_PID` is encoded in lots of places and GDB wants the actual PID.

When attaching to a new process, GDB sends `vKill` to the previous process which the target can use to stop debugging the previous thread.

**Comment by daniel5151** [comment]
> When attaching to a new process, GDB sends `vKill` to the previous process which the target can use to stop debugging the previous thread.

Oh, interesting! After doing a bit of reading, it might be the case that unless you specifically spin up a new "inferior" prior to doing a vAttach, GDB assumes you're all done with the last process and detaches from it...

I've never actually played around with the semantics of multi-process debugging in GDB myself, but now you've got me interested to poke around and see what the different behaviors are (i.e: get a little `gdb` and `gdbserver` playground going on my machine that I can test some stuff in)

* * *

Of course, that's neither here nor there wrt. how `gdbstub` should handle proper multiprocess debugging.

The changes in #129 will likely unblock your use-case of "debug a single process at a time, with the ability to jump between which process is being debugged", but it's not tenable for the sort of true "multiple attached processes" type debugging that this tracking issue encompasses.

**Comment by xobs** [comment]
With the recent merge, I've created https://github.com/betrusted-io/xous-core/pull/360 which implements full support for GDB with Xous, including debugging processes in a live device. Switching processes is now possible, though without `info os processes` the user must know what process ID corresponds to which process.

As you note, it's not true multiprocess support, but I think it's still pretty neat regardless!

**Comment by daniel5151** [comment]
Yep, it's a huge first step towards getting "true" multiprocess support in `gdbstub` - thanks again for your help in implementing + testing things!

Of course, it'd be great to land support for `info os processes` as well (i.e: by responding to `qXfer:osdata:read:processes` packets), but I totally understand if that's not something you're interested in contributing at this time (given that you've got your custom `monitor` command that reports active processes instead).

**Comment by jonathanzetier** [comment]
I'm starting to take a stab at this, using the approach you outline (basically adding a new `MultiProcessBase` that's almost identical to `MultiThreadBase`, but with methods that take a `pid` in addition to a `tid`).

From your comment here, https://github.com/daniel5151/gdbstub/pull/129#discussion_r1134195303, am I correct in assuming that with proper multiprocess support, you're okay with moving active PID management out of the `Target` implementations (via the `attach` and `query_if_attached` methods in the `ExtendedMode` trait and `current_active_pid()` in the `CurrentActivePid` trait)?  Currently the way the code is structured, `current_active_pid` seems to be better described as `last_attached_pid`, which works fine for what gdbstub supports now, but seems like it would lead to confusion in truly multiprocess environments.  Especially given that, unsurprisingly, the vanilla `gdb` only sends a `vAttach`  message on the initial attachment.

It would probably make sense to still have an `attach` method, but rather than relying on targets to properly hold onto the PID, it could be changed (or something similar introduced, to avoid breaking changes) to return information that would be returned to the client instead of `report_reasonable_stop_reason`.

Here's the general shape of how vanilla gdb/gdbserver communicate when attaching to two processes, `A` and `B`:

```
Client sends `vAttach;<pidA>`
Server sends `T0006:...;thread:p<pidA>.<tidA>`
Client sends `qC`
Server sends `QCp<pidA>.<tidA>`
Client sends `Hgp<pidA>.<tidA>`
Server sends `OK`
Client sends `vAttach;<pidB>`
Server sends `T0006:...;thread:p<pidB>.<tidB>`
Client sends `Hgp0.0`
Server sends `OK`
Client sends `qC`
Server sends `QCp<pidB>.<tidB>`
Client sends `Hgp<pidB>.<tidB>`
```

I observed this using `gdbserver --multi 127.0.0.1:<port>` (`GNU gdbserver (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1)`), and doing the following in `gdb` (same version)

```
target extended-remote 127.0.0.1:<port>
attach <pidA>
add-inferior
inferior 2
attach <pidB>
```

Unless I'm missing something, GDB's docs are kind of vague about what the "current thread ID" means (returned by the `qC` command), and at least in this basic example it seems to only be used after a `vAttach` message, so keeping the existing `last_attached_pid` behavior might actually be fine...but it would be nice to avoid having `Target` implementations worrying about that and breaking if we find out out it needs to be changed in the future.

The only other place I could find that uses the `current_active_pid` method seems to be in the `qfThreadInfo` response, which doesn't get sent by default (my gdb client prefers sending `qXfer:threads:read`; I had to use `set remote threads-packet off` to get gdb to send a `qfThreadInfo` packet), and found that it actually returned threads for both `pidA` and `pidB`, so we'd really want to keep track of a list of attached pids, which would also be needed to implement `qAttached` outside of `Target` implementations too.

Apologies for the rambling. This is a long way of saying "I think we should move the responsibility of keeping track of attached pids outside of the `Target` implementations and into `GdbStubImpl`", but just wanted to make sure I'm not missing anything, assuming it can be done with minimal API breakage and minimal (ideally no) code size increase.

**Comment by daniel5151** [comment]
Hey @jonathanzetier! Glad to see you're thinking about the design space here. As expected, there's plenty of nuance that we'll need to iron out here, hah.

Before addressing some of your questions / giving some of my own thoughts, quick bit of book-keeping: given that we're gonna be making breaking changes, I've pre-emptively updated the `dev/0.8` branch to match master, so any PRs you whip up should be sent to that branch.

> From your comment here, [#129 (comment)](https://github.com/daniel5151/gdbstub/pull/129#discussion_r1134195303), am I correct in assuming that with proper multiprocess support, you're okay with moving active PID management out of the `Target` implementations

Indeed you are! As part of the breaking changes, I'd like to see these sort of "useful hax" (as introduced in #129, #133) to go away in favor of gating multiprocess support behind `MultiProcessBase` and friends, and tracking the active PID via a new `current_pid: Pid` entry on `GdbStubImpl` (which would subsume instances of `FAKE_PID` / `current_active_pid()` that currently exist around the tree).

Yes, this would mean that implementations that don't use multi-process support will end up paying for an extra usize + supporting code... but I'm not gonna sweat much over that. Even more generally - even if adding multi-process support will result in a bigger than expected binary size bump... I think having this feature in-tree is more important than keeping things crunched down (in much the same way that having multi-thread support in-tree results in larger bins, even on single-threaded targets).

> It would probably make sense to still have an `attach` method, but rather than relying on targets to properly hold onto the PID, it could be changed (or something similar introduced, to avoid breaking changes) to return information that would be returned to the client instead of `report_reasonable_stop_reason`.

Honestly, this `report_reasonable_stop_reason` thing is something we should untangle as part of this work. It's a rare case where `gdbstub` _doesn't_ offer the end user flexibility on how to respond to something, and there have def been cases in the past where folks have been confused about it's semantics.

Here's an idea: what if we added a new `GdbStubStateMachine::Attached` state? i.e: when `gdbstub` receives a `vRun`, `vAttach`, or `?` method - we punt execution back to the user, and have 'em `report_stop` whatever they want? After all - even single/multi threaded targets have an initial `attach` event, hah.

I'd imagine we'd still want to have the `attach()` and `run()` Target methods (invoked prior to the transition to the `Attached` state, in order to communicate intent to the Target).

Alongside this new state transition, it might be worth including a new `Target::set_reasonable_stop_reason() -> Option<SingleThreadStopReason>` method that users could opt-in/out of if they don't care about handling this detail, and instruct `gdbstub` to respond with some fixed value (albeit with the tid/pid values overwritten based on the tracked current tid/pid). Admittedly, I'm not totally sure this is a good idea... maybe we can play around with it during the implementation phase.

> Unless I'm missing something, GDB's docs are kind of vague about what the "current thread ID" means

Yeah... the GDB RSP do be like that. I suggest poking around the gdb source code itself to glean what the intended plumbing here should be.

My interpretation of this packet has been that it's tied to the infra that drives `self.current_mem_tid` and `self.current_resume_tid`. These values are sometimes updated explicitly (i.e: via the `H` packet), but often times, they are tracked somewhat implicitly (e.g: in response to a stop reason - see code in `write_stop_common`). Presumably, once we start tacking the last active PID, we can poke around what parts of the codebase are currently setting those tid vars, and update them to also update the pid.

> The only other place I could find that uses the `current_active_pid` method seems to be in the `qfThreadInfo` response,

This, yet again, seems like it should be ez to swap out for `self.current_pid`, once it exists.

* * *

Some other thoughts:

- I'm pretty sure we're on the same page here, but just in case: While it seems correct for `gdbstub` to handle tracking `current_tid`, I would be wary of having `gdbstub` track any kind of list of attached processes/threads (which can be arbitrarily large). So while there'll always be some pid/tid tracking outside of `gdbstub` / on the `Target` - it should only be relegated to a few commands (i.e: `qfThreadInfo` / `qXfer:threads:read` and whatnot).

- One important thing to note about `ExtendedMode` is that I wrote the implementation of this IDET relatively early in `gdbstub`'s development, and honestly... the trait doesn't really match the design ethos of the rest of `gdbstub`. There's no strict protocol-level reason to require supporting the `{vRun, vAttach, k, R}` packets as a single indivisible bundle. It seems likely that as part of this work - we'd want to chunk-up `ExtendedMode` into yet more sub-IDETs, in order to give users flexibility wrt lighting up features on an as-needed basis.

-  I wonder whether `MultiProcessBase` might want to take a trait-level dependency on `ExtendedMode`? I don't actually think there's any reason this needs to be the case (at least from a protocol-level POV), but it kinda feels like all the "interesting" bits related to multiprocess support are gated behind ExtendedMode? Maybe the answer will reveal itself during the implementation phase...

- `BaseStopReason` will need a breaking change in order to stuff `pid` info into its various variants. Past-me probably should've had the forethought to include a `Pid` generic alongside the `Tid` generic... but ah well. No biggie.

- We'll prob need to audit the litany of IDETs already implemented in `gdbstub` to make sure that any IDETs that take a `tid: Tid` are updated to also take a `pid: Pid`. I'm not totally what the best approach will be to handle single/multi-threaded variants of all those IDETs (maybe it's fine to just unconditionally pass pid/tid values, even if they're "dummy" values? I think we might even do that today lol). Anyways - just another thing to keep an eye out for.

If we can get this working, I'd be more than happy to immediately cut a `0.8` release of `gdbstub`. This is something I've wanted to see implemented in `gdbstub` for a _long_ time, and I'm sure plenty of people out there would be more than happy to go through the breaking API churn in order to have proper polished mutli-process debugging support.

**Comment by jonathanzetier** [comment]
Great, glad we're on the same page about tossing out some of the workaround traits, and also relieved that an extra Pid-related `usize` or two for targets without Pids isn't a major concern!

>Honestly, this report_reasonable_stop_reason thing is something we should untangle as part of this work. It's a rare case where gdbstub doesn't offer the end user flexibility on how to respond to something, and there have def been cases in the past where folks have been confused about it's semantics.

I just created #194, and of course as soon as I re-read your paragraph, and I seem to have accidentally misread it as "untangle FROM this work" rather than what you actually said, which was "untangle AS PART of this work".  If we can help it, I feel like the PR for adding multiprocess support is going to be large enough that it might be easier to do these in independent PRs; in my own work I personally have a hard time reviewing PRs past a certain size, but if you're comfortable with reviewing both, then you or I can close #194 and I can lump it into the future PR that gets merged.

>I'm pretty sure we're on the same page here, but just in case: While it seems correct for gdbstub to handle tracking current_tid, I would be wary of having gdbstub track any kind of list of attached processes/threads (which can be arbitrarily large). So while there'll always be some pid/tid tracking outside of gdbstub / on the Target - it should only be relegated to a few commands (i.e: qfThreadInfo / qXfer:threads:read and whatnot).

I definitely agree that gdbstub shouldn't track attached threads (we can get them from  `list_active_threads`), but I'm less sure about attached processes.  To be clear, I'm not talking about attach*able* processes (I haven't figured out how to get gdbserver give me a process list, when I experiment I've been getting pids out-of-band with `ps`), but processes that the user has run an `attach` to, in separate inferiors, which in practice seems like it would be a small number (unless there's automation).  

When I was running experiments with the vanilla `gdbserver`/`gdb`, doing `info threads` displays threads from *all* attached processes (whether it's queried from `qfThreadInfo` or `qXfer:threads:read`), so to work properly, *someone* (`GdbStubImpl` or the target) has to keep track of all the processes that get attached.  Relying on the target to do that correctly seems like it would result in a bunch of duplicated code across codebases, which in practice should be pretty simple (append it to a `Vec` if you can use `alloc`, and probably some kind of fixed-size array for `!no_std`) but still just another thing to ask target implementations to do.  Not to mention that we'd also need to ask targets to remove detached processes too...so my thoughts are that using a `Vec`/fixed-size list (for `std`/`no_std`, respectively) for attached processes in `GdbStubImpl` might be preferable here, but I haven't fully thought this through/actually fully gotten this working yet.

Regarding your last points about adding pids to IDETS/the `BaseStopReason`, something I did for single register access ops, which are generic on just a `Tid`, was define a new type, `ExtendedThreadId` that was just a `struct` with a `pid` and `tid` fields, and implement `IsValidTid` for it.   This approach might be useful elsewhere: traits currently have a generic parameter called `Tid` could rename the generic parameter to `ThreadIdType`, which would be `()` for single-thread, `Tid` for multi-thread, and `ExtendedThreadId` for multi-process.  At the end of the day, the `pX.Y` syntax is described by the GDB RSP docs as a type of thread ID.  Of course I'm open to different names, and also I'm still very much in the exploratory phase right now and still don't have it compiling yet, even with liberal `todo!()` usage, so for all I know this might back me into a corner.  I'd *hope* that this wouldn't blow up the code size, hoping the compiler is smart enough to recognize that the `ExtendedThreadId` is small enough that it would pass the pid/tid in the same amount of space as it would pass two separate pid/tid arguments to functions.

> and I'm sure plenty of people out there would be more than happy to go through the breaking API churn

Count me as one of those people, this is something we're motivated to support :-)

**Comment by daniel5151** [comment]
> If we can help it, I feel like the PR for adding multiprocess support is going to be large enough that it might be easier to do these in independent PRs

Sure, we can split up this work into as many PRs as you'd like. When I said "as part of this work", I meant "this work" in the sense of the broader multiprocess support workstream.

FWIW though, I'm not actually opposed to reviewing big PRs, as long as they are semantically indivisible. Artificially slicing big semantically indivisible PRs into smaller ones to make things "easier to review" is a bit of a pet-peeve of mine, hah.

Separate from PR sizing points - I do want to take a sec and make sure that we're on the same page wrt. `report_reasonable_stop_reason` needing to go away. I know you originally suggested tweaking the attach() method to just include extra context to report a reasonable stop reason, but that's not an option I'd like to go with. The `Target` API hasn't included any form of stop-reason handing since the initial refactor to a state machine model, so re-introducing those kinds of semantics isn't a direction we'd want to go in.

* * *

> I definitely agree that gdbstub shouldn't track attached threads (we can get them from `list_active_threads`), but I'm less sure about attached processes.

A few things to consider.

First, as I've alluded to in my first comment - GDB RSP packets tend to fall into one of two camps:
- Explicitly specifying a target thread-id as part of an operation (which most 'modern' packets do)
- Relying on the implicit global "active thread-id" when performing an operation (which many 'older' packets do)

Notably, in either mode - there isn't really a protocol-level need for the GDB server to maintain a list of attached threads/processes. Indeed, the only entity that really needs to maintain that sort of list is the GDB client itself, and the way it "seeds" that mapping is via a handful of specialized query packets (`qfThreadInfo`, `qXfer:threads:read`) that fetch info about available processes / threads.

While you're not wrong insofar that `gdbstub` _could_ maintain an internal representation of what Target processes/threads are available... my claim is that this is a lot of finicky book-keeping for not too much gain (one or two packets?), when it makes far more sense for the Target itself to report what processes/threads it is working with.

Moreover, thinking about it a bit further... I'm not sure that it's even _possible_ for `gdbstub` to maintain a bulletproof internal representation of what threads / processes are in the debugged process? After all, once you resume execution, `gdbstub` gets out of the way and the resulting program can spin up threads / child processes however it likes. Targets are not obligated to report stop-reasons when new threads/pids come into existence, so once you resume, the GDB client undoubtedly will want to accurately rebuild it's view of the world, which necessitates communicating with the underlying `Target`.

It seems that, no matter how you slice it, you'll still end up needing a Target handler to enumerate available threads/pids... and if you're doing that, why have a separate internal mapping at all?

* * *

As for the `BaseStopReason` approach - that _sounds_ plausibly correct? In fact, now that you've pointed out that approach, I think  that might even be what past-me might've intended (at least, based on how confident he sounded in the OP wrt. "aside from the missing #[non_exhaustive], this could've been entirely non-breaking...")

I think it'd hard to reason about without seeing some code, so looking forward to seeing what you come up with!

**Comment by jonathanzetier** [comment]
I almost have a `todo!`-free implementation that builds (minus the additional armv4t implementation, minus actually testing anything yet, so...still some ways to go).  But the last `todo!` I got to involves revisiting the point about the list of attached processes.

I think we might be talking about different things.  I 100% agree with you that having gdbstub keep track of what Target processes/threads are **available to attach to** (`info os processes` in gdb) is a flat out bad idea.  What I'm proposing is gdbstub keeping track of only the processes (not threads) that a client **has attached to** (akin to `info inferiors` in gdb).

The last part I've been working on is the resume logic, and this is another place that would benefit from a list of attached processes: if a `vCont` packet is sent to resume all of its attached processes (gdb has to be coerced to send this this using `set schedule-multiple` before continuing), the only packet that gets sent is `vCont;c`, with no thread ID, which again, isn't being sent to every process on the target, just the ones that the debugger is attached to.  From source code, it seems that gdbserver accomplishes this by iterating through a list of attached inferiors that it keeps track of, and this approach also seems the most natural way to accomplish this in gdbstub.

I'll preface this by saying I haven't worked with gdb as long as you have, so I'm not as aware of all the different edge cases that need to be considered, but it seems like bookkeeping for attached processes would fairly manageable: when `vAttach` is received, put the pid in the list.  When `D;pid` is received, remove the pid from the list.  If a process dies, remove it from the list (probably? this is something I want to check to see what gdb/gdbserver does).  IIRC there's a user setting in the gdb client to decide whether to attach to newly forked processes, so I'd expect gdbserver to rely on gdb to `vAttach` newly created processes, so I don't think gdbstub would need to do anything special in this case.

If I'm missing an edge case, please let me know so I can start thinking of a different way to implement how to resume all the processes that handles that edge case too, and/or if you have any ideas that might work better!

I also think this would take a fundamentally different approach than #178; for multithreading, "all" the threads includes threads the client hasn't seen or has any concept of, so gdbstub needs help from the target (via the `MultiThreadSchedulerLocking` IDET) to enumerate all the possible threads.  For resuming all processes, "all" the processes means only the processes the client has explicitly attached to, so it doesn't seem like there's anything gdbstub needs from the target in this case.

**Comment by daniel5151** [comment]
> I almost have a `todo!`-free implementation that builds (minus the additional armv4t implementation, minus actually testing anything yet, so...still some ways to go).

Very exciting! 

Indeed, given that this is going to be a major feature addition to `gdbstub`, we def can't skimp on adding in-tree validation for multi-process mode.

wrt. the armv4t implementation - it _should_ be relatively straightforward to port the armv4t_multicore example to implement `MultiProcessBase` by having it spin up new emulator instances on-attach. That seems like a plausibly reasonable testbed for experimentation, testing, and validation?

That said, I'm also open to alternative testing approaches. e.g: maybe it'd be easier to spin up a new example that 'emulates' a trivial subset of whatever isa you want (i.e: an x86 system where RAM is filled with `nop` instructions, a-la the in-tree `example_no_std`, or [daniel5151/gdb-bug-repros](https://github.com/daniel5151/gdb-bug-repros)), or hack together some super-minimal stub that wrangles linux processes. Idk, just spitballing.

* * *

> From source code, it seems that gdbserver accomplishes this by iterating through a list of attached inferiors that it keeps track of, and this approach also seems the most natural way to accomplish this in gdbstub.

But unlike `gdbstub`, `gdbserver` serves double-duty as both a GDB RSP server, as well as being a process-level debugger. As a result, I would be a bit wary of drawing architectural conclusions from its implementation (even if it is the canonical source for RSP packet-level semantics).

Or, put differently - you could just as easily argue that it is the "Target" part of `gdbserver` that manages the list of attached inferiors, whereas the "GdbStub" part of `gdbstub` only queries it when handling packets.

> I'm not as aware of all the different edge cases that need to be considered

I might think of more, but right now, it feels like the most interesting edge case is what happens when the process you're debugging forks off a child process.

In these cases, there does exist [`BaseStopReason::Fork`](https://docs.rs/gdbstub/latest/gdbstub/stub/enum.BaseStopReason.html#variant.Fork) which can be used to report that event... but I don't actually think it's _required_ (from a correctness POV) to send that info to the GDB client on-fork. e.g: in cases where you're debugging a program that forks _a lot_, stopping on each fork seems like it'd tank the debugging experience. I don't actually know, this is just conjecture, so double check me here. Maybe this is the user-setting you're talking about?

Lets assume I'm right and it's optional - what happens when a fork occured, and then at some later point, control is yielded back to the GDB client? Will it query to see if any new processes have appeared, or will it just be blind to that newly launched process? Or, on the flipside, if you do report the fork stop reason - does the GDB client automatically attach to the newly forked process as a inferior being debugged? And will it do so explicitly (by sending a separate vAttach), or implicitly (by virtue of having received a stop reason on the fork event)?

I'll also note that much of this discussion wrt. forking off a new process can be projected downwards the abstraction chain to instead be wrt. thread create/exit events (which, fwiw, are encoded in the GDB RSP when [`QThreadEvents`](https://sourceware.org/gdb/current/onlinedocs/gdb.html/General-Query-Packets.html#QThreadEvents) is enabled, though admittedly - this seems to be related to Non-Stop mode, which is a whole different can of worms...). i.e: getting a clearer understanding of how the GDB client enumerates attached threads could shed light on how to handle attached processes

...though even the statement "how to handle attached processes" is a bit iffy for me, given that - at first blush - it seems to me that outside of a few method (attach, run, etc...), you don't really see many standalone `pid`s referenced in the RSP? Rather, the entire protocol is based on having `pid:tid` tuples as the unit of organization, and operations are typically oriented around selecting/operating-on/reporting-the-state-of a single tuple at a time.

* * *

To be clear - I'm not strictly opposed to having `gdbstub` manage a list of inferior processes... but given that I don't have a hands-on familiarity with multi-process debugging, I'd like to really understand the semantics here before implementing an in-stub abstraction that relies on carefully instrumenting RSP packets in order to ensure we don't accidentally forget to update our internal book-keeping.

Also, even if we do pull it off perfectly... I can't help but wonder if the juice is actually worth the squeeze? After all - it seems to me that any non-trivial `Target` that's going to be wrangling multiple inferiors will already be maintaining its own list of inferiors for other reasons... and if so, having it implement a handler based off that info seems pretty easy (rather than duplicating the info and book-keeping logic inside `gdbstub`?)

* * *

I'll end on a more meta-note regarding how I think of the GDB RSP and `gdbstub`'s development philosophy.

There are two sources of pain / complexity in the GDB RSP: the packet-level _syntax_ / _state machines_ of the protocol, and the higher-level aggregate _semantics_ exposed by the protocol.

When I first started the project, I only really thought about the GDB RSP from the perspective of someone developing an emulator, and attempted to streamline both the syntax _and_ semantics of the RSP into a single "easy to use" API. And indeed, if you look at early releases of `gdbstub`, you'll see that there is a _lot_ more internal bookkeeping and hand-holding inside the stub itself.

Taken the the extreme - the very first release of `gdbstub` (0.1) had a `Target::step()` method (that single-stepped the target) that served as the basis of a whole _bunch_ of protocol extensions that were automatically implemented on the user's behalf (e.g: continue is just "keep stepping", breakpoints are just "step until you hit this PC", etc...). This was a perfect, no compromises API, that maximally reduced boilerplate across different emulator implementations... as long as those emulators were implemented as interpreters where the basic execution unit was a single-step (and nothing else).

As time went on and more folks started taking an interest in the project... it became clear that this sort of "semantic streamlining" was fraught with pitfalls, as each new use-case poked holes in what I thought were "reasonable" bits of book-keeping in `gdbstub`. e.g: turns out there are use-cases for `gdbstub` where you might to, oh, I don't know... continue execution _without_ just constantly single-stepping? I know - pretty wild, hah.

So, as time went on, `gdbstub` shifted away from trying to reduce semantic duplication across target implementations, and focused more on designing ergonomic ways to wrap the obtuse and annoying syntax / state-machines of the GDB RSP into nice, type-safe API that folks could interface with via a family of Rust traits.

As we continue to explore multi-process debugging, keep this philosophy in mind. There are all sorts of wacky use-cases folks may come up with for `gdbstub` in the future, and there are so many "unknown unknwons" wrt. not-yet implemented protocol extensions that may end up poking holes in whatever abstractions we come up with here.

Personally, I'd lean towards a simpler design for this initial implementation. There are already plenty of folks interested multi-process support in `gdbstub` (i.e: folks on this thread, as well as those linking to this issue from other repos), and I'm sure we'll get at least a few additional real-world example of multi-process `Target`s once we land something. And once those implementations exist - it'll be a lot easier to tell whether there's some universal semantic streamlining that `gdbstub` can incorporate into its codebase.

**Comment by jonathanzetier** [comment]
I was also curious how forking is handled, so I did an experiment, and my expectation about the client needing to be the one to vAttach to forked processes turned out to be...drumroll...totally wrong!  It does depend on the value of the `set follow-fork-mode`.  When it's `parent` (the default), the sequence of packets after the debugged process forks is:

```
stub -> client: T05fork:pNEWPID.NEWTID;<some registers>
client -> stub: HgpNEWPID.NEWPID
stub -> client: OK
client -> stub: z0,<address of `_dl_debug_state`>
stub -> client: OK
client -> stub: HgpOLDPID.OLDTID
stub -> client: OK
client -> stub: D;NEWPID
stub -> client: OK
client -> stub: vCont;c;pOLDPID.-1
```

When `follow-fork-mode` is `child`, the client never sends a `vAttach` packet, but the client basically sends all the packets it sends after attaching to a process manually with the `attach` command, so experimentally, it seems that whenever a process forks, the gdb client assumes that the process is attached and that it's up the client to DETACH (depending on the setting of `follow-fork-mode`), so at least wrt forking, there is a consistent behavior we could probably apply at the stub-level.  I haven't had enough time to dig into the source code to figure out whether the implicit attachment is in the `Target`-ish side of gdb's code or not though.

----

All this being said, I see what you're saying about getting painted into a corner by having the stub itself manage attached pids, and I'm really just including the above as documentation to refer back to if this goes back to the drawing board.

I think the question of whether the juice is worth the squeeze is a good point.  I've had some time to think since my last comment, and what might strike a better balance is just being able to retrieve a list (or more likely, `Iterator`) of attached pids from the `Target` (maybe from an IDET?), rather than having gdbstub manage the attached pids all on its own.  This lowers the "squeeze", but keeps most of the "juice" of being able to implement some packets at the stub level.

The packet that motivates me the most is `vCont` when no pid and/or tid is specified (aka resume all the attached processes); there's going to need to be a multiprocess variant of https://github.com/daniel5151/gdbstub/blob/master/src/stub/core_impl/resume.rs#L160, and if the stub doesn't have a way to get the attached processes, it seems like `Target`s would have to implement their own version of that function just to handle this case.  Alternatively, we could handle it at the stub level without and just return an error to the client saying "TODO: we don't handle resuming all attached processes at this time", and punt it to later.

----

I currently have the stub tracking keeping track of pids in my local implementation, so I'll try changing that to just retrieving the pids from the target, finish up the multiprocess `vCont` packet, and get some kind of multiprocess `armv4t` target to test with, and then hopefully have a draft PR to push!  Progress is in fits and starts (my wife and I semi-recently had another kid), but like I think I mentioned in a previous comment, this is something my company is enthusiastic about adding support for, so it's not going to die on the vine, it just might take a little longer to get out the door :-)

**Comment by daniel5151** [comment]
Good to see you're digging into the nuts and bolts of how attach semantics work! I've found that empirical observation of how "official" clients work tends to be the most straightforward way to get a feel for how the protocol works... but don't be afraid of poking around the GDB source itself to validate / verify assumptions as well :)

* * *

Unlike `do_vcont_single_thread` vs `do_vcont_multi_thread` - I _strongly_ suspect `do_vcont_multi_process` will be _almost identical_ to `do_vcont_multi_thread`, albeit with some extra `Pid` wiring (and maybe using `Option<Tid>` to model cases of "resume all threads in a particular process").

i.e: just like multi-threaded debugging today, multi-process debugging will invoke multiple `set_resume_action_*` handlers in sequence, depending on what the `vCont` payload is.

So, even in the example you cite (vCont with no pid/tid), the way I see that being reported to the target is just via a standard `clear_resume_actions()`, followed by a `resume()` call (i.e: without calling any `pid.tid` specific methods like `set_resume_action_continue` or whatnot), exactly as it works in multi-thread debugging today. As per the docs - in the absence of explicit per-thread / per-process resume instructions, the Target would simply resume all threads (and in multi-process, across all procs) its attached to - respecting whatever scheduler locking mode is set, ofc.

So, with that in mind - I still don't really see why the stub would need to track attached `pid`s in any capacity / query the Target for what `pid`s it is tracking\* 😅 

As I've pointed out before, it really feels like any non-trivial multi-process `Target` will need to have _some_ internal representation of what PIDs its attached to... and if it already has that, then it should be trivial to implement the resume flow outlined above without needing gdbstub to explicitly specify a resumption behavior per process?

\*outside of having a multi-proc variant of `list_active_threads` (needed to implement `q{f,s}ThreadInfo`), but even this packet needs to return a list of tids, as opposed to a list of bare pids (sans tids)

* * *

And yeah, no worries on timelines! I'm certainly not in any rush, and my primary goal has always been to land high quality code (even if that means it needs more time to bake).

**Comment by daniel5151** [comment]
Also, tangentially related - I poked and prodded an LLM to help me grok the current state of affairs when it comes to multi-_arch_ multi-process debugging. As with all ~slop~ LLM generated content, take it with a grain of salt... but I'd be inclined to agree with the robot that `gdbstub` won't need to support multiple `Arch` implementations on a single stub.

https://gemini.google.com/share/48ddd9d49135

I also briefly poked and prodded the LLM on the more general topic of the role of `pid`s at the RSP level to vibe-check some of the thoughts I've had on this thread. Again, take whatever the robot says with a grain of salt... but it seems like it did a pretty reasonable job analyzing the situation / the role of `pid`s in the protocol.

**Comment by jonathanzetier** [comment]
Yup, my local implementation of `do_vcont_multi_process` so far is almost identical to the multi threaded version!  What I was hoping to do was something like (please pardon the extremely handwavy pseudo-code):

```rs
match action.kind {
    ...
    VContKind::Continue | VContKind::ContinueWithSig(_) => {
        match action.thread {
            None => {
                for pid in <hypothetical attached pid list method> {
                    ops.set_resume_action_continue(pid, SpecificIdKind::All, signal)
                }
            }
            (pid, SpecificIdKind::All) => {
                ops.set_resume_action_continue(pid, SpecificIdKind::All, signal)
            }
            (pid, tid) => {
                ops.set_resume_action_continue(pid, tid, signal)
            }
        }
        ...   
    }
}
```

That being said, I suppose we could also just as easily pass some kind of `SpecificIdKind::All` for the process in the `None` arm, and leave it up to the implementations, if they're going to keep track of their attached pids.

Speaking of, do you have any thoughts about not using scheduler locking for multiprocessing (the method used in #179)?  It seems like one of the main benefits of that approach was not breaking backwards compatibility, but since multiprocessing doesn't exist there's nothing to break.  It seems like in multiprocess environments, it would be a lot more complicated: suppose there's a `vCont;c:pXpid.Xpid;c:pYpid.-1;c:pZpid.-1` packet; we'd need to pass the fact that process `X` is resuming only one thread, but process `Y` and process `Z` are resuming all of their threads.  If a `Target`'s `set_resume_action_continue` function sees more than one process and doesn't support that, it can just return an error saying it doesn't support that.

Tangentially related: the `vCont` documentation states (emphasis added):
>For each inferior thread, the **leftmost action** with a matching thread-id is applied.

It's currently up to targets to decide how they want to handle misbehaving gdb clients; it seems like it would be really easy for implementations to simply do whatever action was set most recently in the `set_resume_action_XXX` methods; it might be something worth documenting in the `MultiThreadResume` trait that if a `Tid` has already been "visited" by a `set_resume_action_XXX` method, it should be ignored...

----

With retrieving the list of all threads, indeed, we're supposed to return a list of all pids/tids, not just pids.  However, I was assuming that the multiprocess version of `list_active_threads` would take a `pid` argument, so this would also be a fairly simple (again, pardon the pseudocode)

```rs
for each pid in <hypothetical attached pid list method> {
    for each tid in list_active_threads(pid, <a similar function as in the multi-threaded case>);
}
```

**Comment by daniel5151** [comment]
In both of these cases, there seems to be a pretty straightforward reformulation that doesn't rely on `gdbstub` tracking PIDs:

> I suppose we could also just as easily pass some kind of `SpecificIdKind::All` for the process in the `None` arm, and leave it up to the implementations, if they're going to keep track of their attached pids.

Or just make the function signature `set_resume_action_continue(pid: Pid, tid: Option<tid>)`, where `None` indicates all TIDs in the process.

> However, I was assuming that the multiprocess version of `list_active_threads` would take a `pid` argument, so this would also be a fairly simple

I don't think that's necessary? Just update the callback function accept a `pid` alongside each `tid`, and have the implementor manage the nested for loops.

* * *

Also, tangentially related: while reading the current `do_vcont_multi_thread` code, I realized it never actually validates that all actions correspond to the same `pid` lol. I suppose that wouldn't actually come up when using a well behaved GDB client + a well behaved gdbstub multi-thread Target... but it's an ez thing to sanity check, so it might be worth adding.

That's probably something that should be fixed in its own PR though, hah.

(also, see my thoughts wrt. handling misbehaving GDB clients below)

* * *

As for scheduler locking... yes, this was added quite recently. @Satar07 drove the testing / validation of this feature, so maybe they have some thoughts on the design in the context of multi-process debugging (if you see this, feel free to chime in on this thread!).

That said, I'm not totally sure I understand your comments wrt. back-compat, and the issues around the multi-pid example you showed.

wrt. back compat: There was a bit of tension wrt. how to handle cases where a client enables scheduler locking without the stub supporting it, and we resolved that tension by adding a descriptive error path to notify the end user that they're missing support in their stub (rather than trying to "limp along" with incorrect semantics). I would imagine this flow would still need to be there in the case of multi-process mode - nothing really changes?

ofc, it'll be up to you to understand how the `vCont` semantics change when scheduler locking is on/off when debugging multiple processes, and make sure `gdbstub` is detecting it correctly. It's lamentable that scheduler locking needs to be "inferred" as opposed to there being an explicit packet toggling it on or off... but hey - that's the GDB RSP for ya.

wrt. the multi-pid example you showed: would this not simply translate into the following Target calls?

```rust
set_resume_action_continue(pidX, Some(tidX));
set_resume_action_continue(pidY, None);
set_resume_action_continue(pidZ, None);
// once resume() is called, depending on scheduler-locking is detected,
// unreferenced PIDs are either kept paused / implicitly resume
resume(); 
```

That seems pretty straightforward to handle, no?

* * *

> Tangentially related: the `vCont` documentation states (emphasis added):
> 
> > For each inferior thread, the **leftmost action** with a matching thread-id is applied.
> 
> It's currently up to targets to decide how they want to handle misbehaving gdb clients; it seems like it would be really easy for implementations to simply do whatever action was set most recently in the `set_resume_action_XXX` methods; it might be something worth documenting in the `MultiThreadResume` trait that if a `Tid` has already been "visited" by a `set_resume_action_XXX` method, it should be ignored...

Excellent observation! I wonder if this was added to the docs at some point after I implemented stuff in `gdbstub`, or if I just missed this blurb. In either case - this is def something we should do something about.

IMO, the fact GDB calls out this leftmost action rule implies that they didn't want to be firm about requiring stubs to track "visited" states for specific thread-ids, and expected implementations to have these kinds of "override / refine" semantics.

e.g: based on this reading, a vCont packet like `vCont;s:X.X;c:X.X;cX.-1` would be legal (albeit a bit weird), and instruct the stub to single-step `X.X`, and continue all other threads.

I think the fix here is two fold:

1. in the code: we update the `ActionsBuf` iterator to return entries right-to-left, so that Target impls only need to care about the last call related to a specific PID 
2. in the docs: we make it (even more) clear that `Target` implementations MUST avoid resuming execution until the `resume()` method has been called, specifically because intermediate calls to `set_resume_action_*` might overwrite / refine the resumption semantics of a particular PID

Technically, changing the iteration order is optional if the docs state that only the first call related to a specific PID matters... but that seems less aesthetically pleasing (at least to my API design sensibilities).

* * *

Another bit of design food for thought: there are many packets (not just vCont) that rely on GDB RSP clients sending over semantically reasonable payloads, and indeed, if a misbehaving client decided to do something unexpected, `gdbstub` and user's `Target` implementations end up getting confused / crashing outright. e.g: see the [`Flash` IDET](https://docs.rs/gdbstub/latest/gdbstub/target/ext/flash/trait.Flash.html).

But honestly, that's OK!

`gdbstub` is _not_ designed to be maximally defensive against misbehaving clients, as in practice, there are only ever 2 clients that ever connect to `gdbstub` - GDB, and LLDB. `gdbstub` is designed under the assumption that these clients are well behaved, and don't have bugs related to RSP semantics. 

Sure, we _could_ add more code into `gdbstub` to detect violations of multi-packet RSP state machine semantics (e.g: behind a feature flag, and/or in debug builds)... but at this stage in development, `gdbstub` pretty much assumes that GDB RSP clients are "gonna do the right thing", and in turn - we document those semantics in the APIs that rely on them.

Now, on the flip-side, `gdbstub` _does_ take extra care to design APIs that make it nearly impossible for `Target` implementors / library users to inadvertently mess up RSP semantics, including adding extra runtime / compile time guard rails to warn when the client attempted to use a unsupported feature (when it's easy). But of course, at a certain point, there will be parts of a `gdbstub` integration that users are simply required to "get right", and so the best we can do is clearly document the semantics we are aware of and require implementations to conform to.

**Comment by jonathanzetier** [comment]
As an aside, I do want to thank you for engaging with me here; I think we just have different tastes when it comes to API design, so hopefully ironing some of these out at the beginning helps make the process smoother for both of us :-)

Backing up wrt resuming: my understanding of the need to introduce `MultiThreadSchedulerLocking` trait in #179 was to avoid breaking existing multithreaded target implementations, and might not be needed in the more "blank state" of multiprocessing, where there aren't any existing targets that might break.  Unless I'm missing something, it seems like the same thing could be accomplished more cleanly (if it was okay to break existing implementations) by:
* Changing `set_resume_action_continue` to take an `Option<Tid>` (and in the multi-process case, an `Option<Pid>` too; my thought with using 
* Replacing the default behavior that "by default, any thread that wasn't explicitly resumed by a `set_resume_action_XXX` method should be resumed as though it was resumed with `set_resume_action_continue`" with "by default, any thread that never had a `set_resume_action_XXX` method called on it stays the same".
* If, when the `resume` method is called, it's being asked to only continue one thread at a time (leaving all others stopped), then the target returns an error saying it can't do that, the same way it might return an error if the client asked to resume a process it never attached to, or a process/thread that doesn't exist.

Please let me know if I'm misunderstanding something; in particular I'm wondering how load-bearing is the assumption that threads are implicitly resumed by `set_resume_action_continue` if they weren't specified in the `vCont` message?  If this is a valid approach, would it be worth changing the multithreaded behavior in v0.8 to use it?  If it's not worth breaking the multithreaded API, is it more important for the multiprocess to look like the multithreaded API or is it okay to have more meaningful differences between the two?

**Comment by daniel5151** [comment]
Of course! I'm more than happy to engage in this sort of back and forth. Admittedly, it _would_ be nice to gradually migrate this back-and-forth to a WIP PR... but this is OK too - I'm happy to work either way for the time being.

* * *

Ahhhh, okay - I think I see what you mean now. Yeah, I totally didn't fully page back in the context wrt. scheduler-locking stuff, lol

Yes, I now see that switching the default behavior on `resume()` to be "any thread that never had a `set_resume_action_XXX` method called on it stays the same" is def the right call, and is closer to the documented `vCont` semantics. And indeed, once we do that - we'll no longer need that separate explicit scheduler locking IDET anymore, and can document `scheduler-locking` semantics on the `resume()` method's docs, punting support for those semantics to the target itself.

As part of moving to that model in multi-process, lets also fixup multi-thread to use those semantics as well.

If you happen to be planning to send multiple smaller PRs here - this is def the sort of change that could be landed first (tweaking existing multi-thread semantics), and then built upon in a subsequent PR (introducing multi-process semantics).

> and in the multi-process case, an `Option<Pid>` too

Indeed... this is a bit of an unfortunate wrinkle.

Because of the whole "An action with no thread-id matches all threads" case in vCont, we'll need `set_resume_action_*` handlers to support handling this "wildcard" case.

In terms of API design, this means that the function signature of `(pid: Option<Pid>, tid: Option<Tid>)` isn't quite right, as it requires users to handle the nonesensical scenario of being given a wildcard PID, but a specific TID.

Instead, the signature would have to be structured as `(thread_id: ThreadId)` where `enum ThreadId { All, AllInPid(Pid), Specific(Pid, Tid) }`. Enum name / variant names subject to bikeshedding, ofc.

Sidenote: the current internal `SpecificThreadId` type might need some rethinking as well, given that it suffers from the same issue of being able to encode the bogus "wildcard pid, specific tid" variant. That said, I haven't scanned the code too deeply... it might still be a reasonable intermediate abstraction, depending on at what point it's being parsed / validated.
